### ISSUE TITLE
Potential performance improvement, filter instead of trim: true

### ISSUE THREAD

**Issue by AleksandarFilipov** [elixir-lang/elixir#11398]
Dear team who enables me (us) to do magical thingies:

While trying to optimize some performance sensitive code, I made the following discovery:
```elixir
String.split(@some_string, " ") |> Enum.filter(fn(x) -> x != "" end)

String.split(@some_string, " ", trim: true) end)
```
The prior solution solution with `filter`, ouperforms `trim: true` but at least factor **2**

Attaching some code:
```
  @some_string "1632767814.9532351 00 0  0 1  0 0 0 0 0 0 0 0 0 0  00  00 0 00 00 00 00 00 0e 01 14 3c 3c 96 20 99 ba 1e 10 80 93 a7 1d 00 00 01 00 00 00 00 00 00 00 00 00 00 00 0  0  0"
  def bench_string_split_trim(n) do

    fun1 = fn ->
      Enum.map(1..n, fn(_x) ->
      String.split(@some_string, " ", trim: true) end)
    end

    fun2 = fn ->
      Enum.map(1..n, fn(_x) ->
      String.split(@some_string, " ") |> Enum.filter(fn(x) -> x != "" end) end)
    end

    Benchee.run(
      %{
        "vanilla" => fun1,
        "filter" => fun2,
      },
      formatters: [
        {Benchee.Formatters.HTML, file: "samples_output/my.html"},
        Benchee.Formatters.Console
      ]
    )
  end
```  
 
I wanted to bring this to your attention, possibly there is something useful here.

**Comment by josevalim** [comment]
Can you please include your Erlang, Elixir version, and benchmark results?

**Comment by AleksandarFilipov** [comment]
yeps - should have been in the original issue...
```
Elixir: 1.11.4
Erlang: 23.2.6
Operating system: macOS
Available memory: 8 GB
CPU Information: Intel(R) Core(TM) i7-3635QM CPU @ 2.40GHz
Number of Available Cores: 8
```
![newplot (1)](https://user-images.githubusercontent.com/24603786/142216372-844fde1d-1ca4-408b-98ed-b58f5c3d6c01.png)
![newplot](https://user-images.githubusercontent.com/24603786/142216401-9c151701-6f5d-4c21-8e6e-8c0d0afeba98.png)

**Comment by josevalim** [comment]
Thanks! This has been fixed since then (potentially on v1.12):

```
Operating System: macOS
CPU Information: Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
Number of Available Cores: 8
Available memory: 16 GB
Elixir 1.13.0-rc.1
Erlang 24.0

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 5 s
memory time: 0 ns
parallel: 1
inputs: none specified
Estimated total run time: 14 s

Benchmarking filter...
Benchmarking vanilla...

Name              ips        average  deviation         median         99th %
vanilla      484.68 K        2.06 μs  ±1010.33%        1.99 μs        3.99 μs
filter       254.78 K        3.92 μs   ±323.51%        3.99 μs       14.99 μs

Comparison:
vanilla      484.68 K
filter       254.78 K - 1.90x slower +1.86 μs
```

**Comment by AleksandarFilipov** [comment]
waoo, that's what I would call express delivery :). Not only fixed by also significantly faster the "other way around", as one would wish.

Thanks @josevalim
